src/app/3d/qgs3daxis.cpp

```diff
@@ -0,0 +1,643 @@
+#include <Qt3DExtras/QPhongMaterial>
+#include <Qt3DRender/qcameralens.h>
+#include <Qt3DRender/QCameraSelector>
+#include <Qt3DRender/QClearBuffers>
+#include <Qt3DRender/QLayer>
+#include <Qt3DRender/QLayerFilter>
+#include <Qt3DRender/QPointLight>
+#include<ctime>
+
+#include <Qt3DRender/QAttribute>
+
+Qgs3DAxis::Qgs3DAxis( Qt3DExtras::Qt3DWindow *parentWindow, Qt3DCore::QEntity *parent3DScene, QgsCameraController *cameraCtrl, const Qgs3DMapSettings *map )
+  : QObject( parentWindow ), mParentWindow( parentWindow ), mParentCamera( cameraCtrl->camera() ),
+    mCrs( map->crs() )
+{
+  mAxisViewport = constructAxisViewport( parent3DScene );
+  mAxisViewport->setParent( mParentWindow->activeFrameGraph() );
+
+  mTwoDLabelViewport = constructLabelViewport( parent3DScene, QRectF( 0.0f, 0.0f, 1.0f, 1.0f ) );
+  mTwoDLabelViewport->setParent( mParentWindow->activeFrameGraph() );
+
+  connect( cameraCtrl, &QgsCameraController::cameraChanged, this, &Qgs3DAxis::updateCamera );
+  connect( mParentWindow, &Qt3DExtras::Qt3DWindow::widthChanged, this, &Qgs3DAxis::updateAxisViewportSize );
+  connect( mParentWindow, &Qt3DExtras::Qt3DWindow::heightChanged, this, &Qgs3DAxis::updateAxisViewportSize );
+
+  updateAxisViewportSize();
+
+  createAxisScene();
+}
+
+Qt3DRender::QViewport *Qgs3DAxis::constructAxisViewport( Qt3DCore::QEntity *parent3DScene )
+{
+  auto axisViewport = new Qt3DRender::QViewport;
+  // size will be set later
+
+  mAxisSceneEntity = new Qt3DCore::QEntity;
+  mAxisSceneEntity->setParent( parent3DScene );
+
+  mAxisCamera = new Qt3DRender::QCamera;
+  mAxisCamera->setParent( mAxisSceneEntity );
+  mAxisCamera->setProjectionType( mParentCamera->projectionType() );
+  // always ortho for now as it seem more "accurate"
+  if ( true || mParentCamera->projectionType() == Qt3DRender::QCameraLens::ProjectionType::OrthographicProjection )
+    mAxisCamera->lens()->setOrthographicProjection(
+      -mAxisViewportSize / 2.0f, mAxisViewportSize / 2.0f,
+      -mAxisViewportSize / 2.0f, mAxisViewportSize / 2.0f,
+      -100.0f, 500.0f );
+  else
+    mAxisCamera->lens()->setPerspectiveProjection(
+      mParentCamera->lens()->fieldOfView(), mParentCamera->lens()->aspectRatio(),
+      1.0f, -250.0f );
+
+  mAxisCamera->setUpVector( QVector3D( 0.0f, 0.0f, 1.0f ) );
+  mAxisCamera->setViewCenter( QVector3D( 0.0f, 0.0f, 0.0f ) );
+  // position will be set later
+
+  auto axisLayer = new Qt3DRender::QLayer;
+  axisLayer->setRecursive( true );
+  mAxisSceneEntity->addComponent( axisLayer );
+
+  auto axisLayerFilter = new Qt3DRender::QLayerFilter( axisViewport );
+  axisLayerFilter->addLayer( axisLayer );
+
+  auto axisCameraSelector = new Qt3DRender::QCameraSelector;
+  axisCameraSelector->setParent( axisLayerFilter );
+  axisCameraSelector->setCamera( mAxisCamera );
+
+  auto clearBuffers = new Qt3DRender::QClearBuffers( axisCameraSelector );
+  clearBuffers->setBuffers( Qt3DRender::QClearBuffers::DepthBuffer );
+
+  return axisViewport;
+}
```

Comment:
Ok, but to avoid to put all the frame graph piece of code into the same class (QgsShadowRenderingFrameGraph) as it may become messy, can we imagine some kind of extension mechanism? Like something with a branch manager into QgsShadowRenderingFrameGraph and a API provided by a base class to register new branch into the manager.

Reply:
Actually I do not think that any extension mechanism should be provided for frame graph management, and everything should be implemented in our existing frame graph class (its name is a bit misleading, we should rename it).

I am open to discussion about this, but my experience so far is that in order to stay sane, it's necessary to keep the frame graph's complexity limited :-)